### PR TITLE
Consider pointer to char - 0 dim as strings

### PR DESCRIPTION
**Pull request by Smit-create**
Part of extending #1148 to LLVM

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -1947,6 +1947,9 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 v_Var, args.p, args.size(), type, ASR::arraystorageType::ColMajor, arr_ref_val);
         } else {
             ASR::ttype_t *v_type = ASRUtils::symbol_type(v);
+            if (ASR::is_a<ASR::Pointer_t>(*v_type)) {
+                v_type = ASR::down_cast<ASR::Pointer_t>(v_type)->m_type;
+            }
```
Comment (Smit-create) [inline src/lfortran/semantics/ast_common_visitor.h:1952]: This change handles `ptr(1:3)` as the `StringSection` where `ptr` is the character pointer. As in the LLVM backend `character_type` is a pointer to the `int8` array, `ptr` should not be a pointer to `character_type` right? I guess `gfortran` also does the same. See the following snippet:

```fortran
program main
    implicit none
    character(len=20), target :: num
    character(len=:), pointer :: ptr
    num = "working"
    ptr => num
    print *, ptr
end program main
```

LFortran output:
```
% lfortran c.f90 
4362780567
```

GFortran:
```
% gfortran c.f90 -o c.out && ./c.out
 working             
```

Moreover, we should also change the LLVM backend to handle `Pointer(Character & n_dims=0)` as a simple string i.e. `character_type`. Is this fine to move ahead?

Comment (czgdp1807) [reply src/lfortran/semantics/ast_common_visitor.h:1952]: I guess this is related to,

https://github.com/lfortran/lfortran/blob/cbcbdcc376ecabb671fde7977a49def799f5f622/src/libasr/pass/print_arr.cpp#L95-L101

So we need to figure out a way depending on each frontend. The output you are getting is the address pointed to by `ptr` and not its contents. Could you please also show the ASR of the Fortran code you posted with `main` and your branch?

Comment (Smit-create) [reply src/lfortran/semantics/ast_common_visitor.h:1952]: Fortran code:
```fortran
program main
    implicit none
    character(len=20), target :: num
    character(len=:), pointer :: ptr
    num = "working"
    ptr => num
    print *, ptr
    print *, ptr(1:4)
end program main
```

This is the ASR:

<details>
<summary> main </summary>
<p>

```console
(TranslationUnit
    (SymbolTable
        1
        {
            main:
                (Program
                    (SymbolTable
                        2
                        {
                            num:
                                (Variable
                                    2
                                    num
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Character 1 20 () [])
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            ptr:
                                (Variable
                                    2
                                    ptr
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Pointer
                                        (Character 1 -2 () [])
                                    )
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    main
                    []
                    [(=
                        (Var 2 num)
                        (StringConstant
                            "working"
                            (Character 1 7 () [])
                        )
                        ()
                    )
                    (=>
                        (Var 2 ptr)
                        (Var 2 num)
                    )
                    (Print
                        ()
                        [(Var 2 ptr)]
                        ()
                        ()
                    )
                    (Print
                        ()
                        [(ArraySection
                            (Var 2 ptr)
                            [((IntegerConstant 1 (Integer 4 []))
                            (IntegerConstant 4 (Integer 4 []))
                            (IntegerConstant 1 (Integer 4 [])))]
                            (Character 1 -2 () [])
                            ()
                        )]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

</p>
</details>

<details>
<summary> this branch </summary>
<p>

```console
(TranslationUnit
    (SymbolTable
        1
        {
            main:
                (Program
                    (SymbolTable
                        2
                        {
                            num:
                                (Variable
                                    2
                                    num
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Character 1 20 () [])
                                    Source
                                    Public
                                    Required
                                    .false.
                                ),
                            ptr:
                                (Variable
                                    2
                                    ptr
                                    []
                                    Local
                                    ()
                                    ()
                                    Default
                                    (Pointer
                                        (Character 1 -2 () [])
                                    )
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    main
                    []
                    [(=
                        (Var 2 num)
                        (StringConstant
                            "working"
                            (Character 1 7 () [])
                        )
                        ()
                    )
                    (=>
                        (Var 2 ptr)
                        (Var 2 num)
                    )
                    (Print
                        ()
                        [(Var 2 ptr)]
                        ()
                        ()
                    )
                    (Print
                        ()
                        [(StringSection
                            (Var 2 ptr)
                            (IntegerConstant 1 (Integer 4 []))
                            (IntegerConstant 4 (Integer 4 []))
                            (IntegerConstant 1 (Integer 4 []))
                            (Character 1 -1 () [])
                            ()
                        )]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

</p>
</details>

Comment (czgdp1807) [reply src/lfortran/semantics/ast_common_visitor.h:1952]: Theoretically the ASR is correct. So, the discrepancy is in the backend which is due to the conflicts between how printing is dealt with in LPython and LFortran.

**Comment by Smit-create** [comment]
How does this look @czgdp1807 @certik?

**Review by czgdp1807 — COMMENTED** [review]
How does this change affects LPython printing?

**Review by certik — APPROVED** [review]
I think the printing is not affected by this PR, only associate.

**Pull request merged**